### PR TITLE
fix(core): don't key a resource under Chrome's abandoned https attempt

### PR DESCRIPTION
**Pull request by rishigupta1599**
## Symptom

A Cypress suite takes three snapshots of the same page URL. The first captures all 14 resources; the other two capture **none** — every sub-resource request goes out over `https://` and fails with `net::ERR_CONNECTION_CLOSED`, and the renderer reports `asset not present`, so the snapshots render completely unstyled. Same host, same CI runner, same `discovery.allowedHostnames` as the 36 snapshots that work.

Across all 39 snapshots of the reported build the correlation is exact: the only three that share a `source-url` are the affected ones, the first of them is fine, and the other two are the only two snapshots in the build with any `ERR_CONNECTION_CLOSED`.

## Root cause

`originURL()` returned `request.redirectChain[0].url`.

Chrome attempts an `http://` main-frame navigation over `https://` first and retries it over `http://` once TLS fails. It surfaces that fallback as a redirect hop, so the successful http request arrives with the *abandoned* https attempt at the head of its redirect chain:

```
Handling request: https://host/page   raw=https://host/page  chain=[]
Handling request: https://host/page   raw=http://host/page   chain=["https://host/page"]
                                      ^ logged and keyed as https://
```

So a plain-http snapshot URL got keyed under an `https://` URL nothing ever referenced. Two consequences:

1. **The snapshot's own root resource was never served.** `parseDomResources` registers it at the declared `http://` URL (`createRootResource` → `normalizeURL`), which never matched the request — `- Serving root resource` does not appear anywhere in the reported 39-snapshot build. Asset discovery silently re-fetched and rendered the **live page** instead of the captured DOM snapshot for every snapshot.

2. **The build-wide discovery cache got poisoned.** The fetched document was saved under the `https://` key. `lookupCacheResource` is keyed by URL alone and shared across the whole build, so on the *next* snapshot of the same URL the https attempt was satisfied by `Fetch.fulfillRequest` from that cache. The upgrade therefore **succeeded**, the document committed over `https://`, and every root-relative sub-resource was requested from a port the origin has no TLS listener on → `ERR_CONNECTION_CLOSED` → nothing captured.

That is why it only ever hit the second and later snapshots of a URL, why it needed a real hostname (Chrome exempts `localhost` and IP literals from HTTPS upgrades, so it does not reproduce locally), and why unrelated snapshots on the identical host were unaffected.

Same family as the earlier method-blind cache issue: the discovery cache is keyed by URL alone, so anything that mis-derives that URL leaks across snapshots.

## Fix

Walk the redirect chain and skip hops that are the abandoned leg of an HTTPS-upgrade fallback — a hop that differs from the one after it only by an `https:` scheme — so a request is keyed by the URL that was really fetched. A real server redirect still reports the URL the document referenced, including one that follows an upgrade fallback.

## Testing

Reproduced against a plain-HTTP origin behind a real hostname (`app.localtest.me`, public DNS → `127.0.0.1`, resolver-mapped so both `:80` and `:443` reach the HTTP-only server), taking two snapshots of one URL.

Before — second snapshot loses everything, matching the reported log:

```
Discovering resources: snap-two
Handling request: https://app.localtest.me/page
- Resource cache hit
Handling request: https://app.localtest.me/style.css
Request failed for https://app.localtest.me/style.css: net::ERR_SSL_PROTOCOL_ERROR
Request failed for https://app.localtest.me/img.svg: net::ERR_SSL_PROTOCOL_ERROR
```

After — the root resource is served (it never was before) and sub-resources resolve over http and hit the cache:

```
Discovering resources: snap-two
Handling request: http://app.localtest.me/page
- Serving root resource
Handling request: http://app.localtest.me/style.css
- Resource cache hit
Handling request: http://app.localtest.me/img.svg
- Resource cache hit
```

The origin also stops being hit for the document at all, since the captured DOM snapshot is now used as intended.

Added six unit tests for `originURL` in `packages/core/test/unit/network.test.js` covering the upgrade fallback, a real server redirect, an upgrade fallback followed by a server redirect, a scheme change with a differing path (not a fallback), a genuine https request, and hash/query normalization.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/core/test/unit/network.test.js [hunk 1]
```diff
@@ -138,6 +138,56 @@ describe('Unit / Network', () => {
   // flattenLookupAddresses — normalizes the dns.lookup callback address argument
   // (Node's http stack passes an array under Happy Eyeballs; others a string) so
   // the direct-fetch choke point can gate on every candidate connection IP.
+  describe('originURL', () => {
```
Comment (rishigupta1599) [inline packages/core/test/unit/network.test.js:138]: **[Low] `flattenLookupAddresses` comment orphaned by this insertion**

The three comment lines directly above this `describe` describe `flattenLookupAddresses`, not `originURL`. Inserting the new block here split that comment from the `describe('flattenLookupAddresses', ...)` it belongs to (now uncommented, further down).

**Suggestion:** move those three lines down to sit immediately above `describe('flattenLookupAddresses', ...)`.

_Reviewer: stack-code-reviewer_

File: packages/core/src/network.js [hunk 2]
```diff
@@ -682,9 +682,29 @@ export function logAssetInstrumentation(log, category, reason, details) {
...
-function originURL(request) {
-  return normalizeURL((request.redirectChain[0] || request).url);
+// True when `from` is `to` with an https:// scheme — the abandoned leg of
+// Chrome's HTTPS-upgrade fallback.
+function isHttpsUpgradeFallback(from, to) {
+  return from.startsWith('https://') && to.startsWith('http://') &&
+    from.slice('https:'.length) === to.slice('http:'.length);
+}
+
+// Returns the URL the document referenced — the head of the redirect chain for
```
Comment (rishigupta1599) [inline packages/core/src/network.js:692]: **[Medium] Same-path https→http server redirects are indistinguishable from the fallback**

`isHttpsUpgradeFallback` is purely structural (same host+path+query, https→http). A genuine 30x redirect from `https://a/x` to `http://a/x` — rare, but produced by misconfigured reverse proxies — would also be skipped, changing which URL that resource is keyed under. That looks like an accepted trade-off, but it is currently implicit.

**Suggestion:** add a sentence to this doc comment stating the ambiguity and why it is acceptable.

_Reviewer: stack-code-reviewer_

**Review by rishigupta1599 — COMMENTED** [review]
Claude Code Review (automated) — 2 inline finding(s). Full report in the PR comment below. Verdict: Passed.

**Comment by rishigupta1599** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2421  •  **Head:** cecec15  •  **Reviewers:** stack-code-reviewer

## Summary
Replaces `originURL`'s `redirectChain[0] || request` with a loop that skips Chrome's abandoned HTTPS-upgrade-fallback hop, so an `http://` snapshot URL is no longer keyed under an `https://` URL nothing referenced — which had hidden the snapshot's own root resource and poisoned the build-wide URL-keyed discovery cache for later snapshots of the same URL.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | No secrets in the diff. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface touched. |
| High | Security | Input validation and sanitization | Pass | URLs parsed via `normalizeURL`/`new URL()`, not string-built. |
| High | Security | No IDOR — resource ownership validated | N/A | No resource-ownership surface. |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL. |
| High | Correctness | Logic is correct, handles edge cases | Pass | Reviewer executed the function against the 6 shipped cases plus default ports, explicit matching non-default ports, differing query string, and scheme case — all correct. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | No new error paths; `chain[i + 1]?.url ?? request.url` guards the last hop. |
| High | Correctness | No race conditions or concurrency issues | Pass | Pure function over a single request object. |
| Medium | Testing | New code has corresponding tests | Pass | 6 unit tests covering fallback, real redirect, fallback+redirect, differing path, genuine https, normalization. |
| Medium | Testing | Error paths and edge cases tested | Fail | Ports, differing query string and scheme case are untested — see finding 2. Verified correct today, but unguarded against future refactors. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full CI green, including `Test @percy/core`. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Loop is bounded by redirect-chain length (0–2 hops typical). |
| Medium | Performance | Long-running tasks use background jobs | N/A | Synchronous helper. |
| Medium | Quality | Follows existing codebase patterns | Pass | Exporting a pure helper for tests matches `pickCookieSession`, `shouldAttachAuth`, `raceWithTimeout`, `resolveDirectFetchMime`, `flattenLookupAddresses` in the same file. |
| Medium | Quality | Changes are focused (single concern) | Pass | Two files, one concern. |
| Low | Quality | Meaningful names, no dead code | Pass | — |
| Low | Quality | Comments explain why, not what | Fail | The `flattenLookupAddresses` comment is now orphaned above the new `describe` — see finding 1. |
| Low | Quality | No unnecessary dependencies added | Pass | None added. |

## Findings

- **File:** `packages/core/test/unit/network.test.js:141`
- **Severity:** Low
- **Reviewer:** stack-code-reviewer
- **Issue:** The comment block describing `flattenLookupAddresses` now sits directly above the newly inserted `describe('originURL', ...)`, while the real `describe('flattenLookupAddresses', ...)` at line 191 is left uncommented. A copy-paste artifact of where the block was inserted; actively misleading read top-to-bottom.
- **Suggestion:** Move that comment down to sit above `describe('flattenLookupAddresses', ...)`.

- **File:** `packages/core/test/unit/network.test.js:141`
- **Severity:** Low
- **Reviewer:** stack-code-reviewer
- **Issue:** No test covers ports (default vs explicit-matching), a differing query string on an otherwise identical path, or scheme case — precisely where `isHttpsUpgradeFallback`'s string-slice comparison is riskiest. The reviewer confirmed all behave correctly today (default ports collapse in `normalizeURL`, a differing query correctly blocks a match), so there is no live bug, but a future refactor could regress one silently.
- **Suggestion:** Add 2–3 cases covering ports and a differing query string.

- **File:** `packages/core/src/network.js:692`
- **Severity:** Medium
- **Reviewer:** stack-code-reviewer
- **Issue:** `isHttpsUpgradeFallback` is purely structural (same host+path+query, https→http). A genuine 30x redirect from `https://a/x` to `http://a/x` — rare, but produced by misconfigured reverse proxies — is indistinguishable from the abandoned fallback leg and would be skipped, changing which URL that resource is keyed under. Presently an implicit assumption rather than a documented one.
- **Suggestion:** Add a sentence to the doc comment acknowledging the ambiguity and why it is an acceptable trade-off.

### Notes verified during review (no action)

- No unintended behaviour change at the other five `originURL` call sites (`network.js:319, 445, 499, 933, 950`). Two are pure logging. The `hostnameMatches` allow/deny checks match on hostname+port only (`packages/client/src/proxy.js:38-58`), never scheme, so the fallback rewrite cannot flip an allowlist outcome. `captureResourceDirectly` now saves worker-script resources under the corrected URL — a positive knock-on consistent with intent.
- The loop is behaviour-preserving for every non-fallback case: empty chain, single real redirect, and multi-hop chains all match the pre-PR `(redirectChain[0] || request).url` semantics exactly.

---

**Verdict: PASS** — core fix is correct and precisely targeted; the one pre-merge-worthy item is the cosmetic misplaced comment.

**Comment by rishigupta1599** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2421  •  **Head:** 8007bf5  •  **Reviewers:** stack-code-reviewer

_Continues [the previous review](https://github.com/percy/cli/pull/2421#issuecomment-5579191526) — changes since `cecec15` (delta)._

## Summary
Delta addresses the three findings from the previous round: documents the same-path https→http ambiguity in `originURL`'s doc comment, adds three `originURL` tests (default-port hop, matching explicit non-default ports, differing query string), and restores the `flattenLookupAddresses` comment to the `describe` it documents.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Comment + test-only delta. |
| High | Security | Authentication/authorization checks present | N/A | No auth surface touched. |
| High | Security | Input validation and sanitization | Pass | Unchanged; URLs still parsed via `normalizeURL`. |
| High | Security | No IDOR — resource ownership validated | N/A | — |
| High | Security | No SQL injection (parameterized queries) | N/A | No SQL. |
| High | Correctness | Logic is correct, handles edge cases | Pass | No production logic changed in this delta; reviewer re-traced `normalizeURL` + `isHttpsUpgradeFallback` against all three new assertions by hand. |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Unchanged. |
| High | Correctness | No race conditions or concurrency issues | Pass | Unchanged. |
| Medium | Testing | New code has corresponding tests | Pass | 9 `originURL` tests now. |
| Medium | Testing | Error paths and edge cases tested | Pass | Ports and query string now covered; scheme case remains unpinned but is unreachable — see remaining finding. |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | Full CI was green on the parent commit; comment relocation is clean with no duplicate left behind. |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | Unchanged. |
| Medium | Performance | Long-running tasks use background jobs | N/A | — |
| Medium | Quality | Follows existing codebase patterns | Pass | — |
| Medium | Quality | Changes are focused (single concern) | Pass | Review follow-ups only. |
| Low | Quality | Meaningful names, no dead code | Pass | — |
| Low | Quality | Comments explain why, not what | Pass | Orphaned comment restored to its `describe`; the new doc sentence states a trade-off, not a restatement. |
| Low | Quality | No unnecessary dependencies added | Pass | None added. |

## Findings

### Resolved since `cecec15`

- ~~`packages/core/test/unit/network.test.js:141` Low — `flattenLookupAddresses` comment orphaned above the new `describe` block~~ **Resolved** — comment moved back above `describe('flattenLookupAddresses', ...)`; reviewer confirmed no duplicate copy left behind.
- ~~`packages/core/src/network.js:692` Medium — heuristic cannot distinguish the Chrome fallback from a genuine same-path https→http redirect, and the assumption was implicit~~ **Resolved** — doc comment now states the trade-off explicitly, and the reviewer judged the scoping accurate (a normal server redirect changes host or path, already covered by the differing-path test).

### Still open

- **File:** `packages/core/test/unit/network.test.js:183`
- **Severity:** Low (unchanged from first sighting in `cecec15`)
- **Reviewer:** stack-code-reviewer
- **Issue:** Scheme case (e.g. `HTTPS://example.test/page`) is still not pinned by a test. In practice this is unreachable: `isHttpsUpgradeFallback` is only ever called on strings already through `normalizeURL`, and the WHATWG `URL` parser lowercases `protocol` and `hostname` regardless of input casing — so a case mismatch cannot reach the literal `startsWith('https://')` check. Carried forward as a documentation gap, not a live defect.
- **Suggestion:** Optional — a single assertion that `normalizeURL` lowercases the scheme, in a `normalizeURL` describe block rather than `originURL`, as insurance against a future refactor that bypasses `normalizeURL`.

### Verified during this round

- All three new assertions were traced by hand against the real `normalizeURL` (`packages/core/src/utils.js:28-31`) and `isHttpsUpgradeFallback` (`packages/core/src/network.js:687-690`), not taken on trust: default-port `:443` collapses via `URL.port === ''`; `:8080` is preserved on both sides and still matches; a differing query correctly blocks the match so the chain head is returned unchanged.
- No new issues introduced by the delta.

---

**Verdict: PASS** — all three prior findings addressed; the single carried-forward Low is a test-documentation gap that cannot manifest at runtime.

**Review by ninadbstack — APPROVED** [review]
